tests.js

```diff
@@ -16,45 +16,167 @@
   }
 
-  list(after, limit) {
-    let begin; let end;
-    if (after) {
-      const afterURL = new URL(after);
-      const afterIndex = this.items.findIndex((item) => {
-        const itemURL = new URL(`${item.protocol}://${this.host}${item.pathname}`);
-        return itemURL.pathname === afterURL.pathname &&
-            itemURL.protocol === afterURL.protocol;
-      });
+  buildEndpoints() {
+    const endpoints = {};
 
-      if (afterIndex === -1) {
-        throw new Error(`${after} not found in test manifest`);
+    for (const [ident, test] of Object.entries(this.tests)) {
+      for (const scope of test.scope) {
+        let endpoint = '';
+        let httpsOnly = false;
+        switch (scope) {
+          case 'Window':
+            endpoint = '/api/interfaces';
+            break;
+          case 'Worker':
+          case 'DedicatedWorker':
+            endpoint = '/api/workerinterfaces';
+            break;
+          case 'ServiceWorker':
+            endpoint = '/api/serviceworkerinterfaces';
+            httpsOnly = true;
+            break;
+          case 'CSS':
+            endpoint = '/css/properties';
+            break;
+        }
+
+        if (endpoint) {
+          if (!(endpoint in endpoints)) {
+            endpoints[endpoint] = {
+              scope: scope,
+              httpsOnly: httpsOnly,
+              entries: []
+            };
+          }
+          if (!(ident in endpoints[endpoint].entries)) {
+            endpoints[endpoint].entries.push(ident);
+          }
+        }
       }
+    }
 
-      begin = afterIndex + 1;
-      if (limit) {
-        end = afterIndex + 1 + limit;
+    return endpoints;
+  }
+
+  listMainEndpoints(urlPrefix = '') {
+    return Object.keys(this.endpoints).map((item) => (
+      ['', `${urlPrefix}${item}`]
+    ));
+  }
+
+  listIndividual(urlPrefix = '') {
+    return Object.keys(this.tests).map((item) => (
+      [item, `${urlPrefix}/${item.replace(/\./g, '/')}`]
+    ));
+  }
+
+  listAllEndpoints(urlPrefix = '') {
+    return [
+      ...this.listMainEndpoints(urlPrefix),
+      ...this.listIndividual(urlPrefix)
+    ];
+  }
+
+  next(after) {
+    const afterURL = new URL(after);
+    if (!this.httpOnly && afterURL.protocol === 'http:') {
+      return `https://${this.host}${afterURL.pathname}`;
+    }
+    const endpoints = this.listMainEndpoints('/tests');
+    const index = endpoints.findIndex((item) => {
+      return item[1] === afterURL.pathname;
+    }) + 1;
+
+    if (index >= endpoints.length) {
+      return null;
+    }
+
+    const endpoint = endpoints[index][1];
+
+    if (this.endpoints[endpoint.replace('/tests', '')].httpsOnly) {
+      const newUrl = `https://${this.host}${endpoint}`;
+      if (this.httpOnly) {
+        // Skip this endpoint and go to the next
+        return this.next(newUrl);
       }
+      return newUrl;
+    }
+
+    return `http://${this.host}${endpoint}`;
+  }
+
+  getTests(endpoint) {
+    let idents = [];
```

Comment:
```suggestion
    let idents;
```

It's assigned in both branches, let's be explicit that this empty array is never used.